c#/Traverse/TraverseFolders/FileSystemVisitor.cs

```diff
@@ -1,72 +1,82 @@
-public delegate bool ReturnedDelegate(string filename, string extention);
+public delegate bool FilterFunctionDelegate(string name);
+
+public delegate void Notify();
+
+public delegate bool MyCustomEvent(string filename,CustomEventArgs eventArgs);
 
 class FileSystemVisitor
 {
+    private readonly string _path;
+    private readonly FilterFunctionDelegate _filterFunc;
+    private readonly CustomEventArgs _eventArgs;
+
+    public event Notify OnStart;
+    public event Notify OnFinish;
+    public event MyCustomEvent OnFileFound;
+
     public FileSystemVisitor(string filename)
     {
-        foreach (string file in Explore(filename)) {
-            Console.WriteLine(file);
-        }
+        _path = filename;
+        _filterFunc = (_) => true;
+        OnStart = () => Console.WriteLine("Process started");
+        OnFinish = () => Console.WriteLine("Process finished");
     }
 
-    public FileSystemVisitor(string filename, ReturnedDelegate delegat)
+    public FileSystemVisitor(string filename, FilterFunctionDelegate delegat)
     {
-        foreach (string file in Explore(filename,delegat)) {
-            Console.WriteLine(file);
-        }
+        _path = filename;
+        _filterFunc = delegat;
+        OnStart = () => Console.WriteLine("Process started");
+        OnFinish = () => Console.WriteLine("Process finished");
+    }
+    
+    public FileSystemVisitor(string filename, FilterFunctionDelegate delegat, Notify started, Notify finished, MyCustomEvent fileFoundEvent)
+    {
+        _path = filename;
+        _filterFunc = delegat;
+        OnStart = started;
+        OnFinish = finished;
+        OnFileFound = fileFoundEvent;
     }
     
-    public IEnumerable<string> Explore(string path, ReturnedDelegate delegat)
+    public IEnumerable<string> Explore(string path)
     {
+        OnStart();
         foreach (var entry in Directory.EnumerateFileSystemEntries(path))
         {
-            if (delegat(entry,"txt"))
+            if (OnFileFound(entry, _eventArgs))
```

Comment:
you can avoid if statement here by modifying eventArgs inside handler and then do break based on flag from args